kratos/utilities/helper_classes_for_constraint_builder.h

```diff
@@ -605,84 +570,137 @@ class ConstraintImposer {
+            // Get the tranformation matrix and constant_vector from the current slave
             global_master_slave_constraint->second->EquationIdsVector(slave_equation_id, master_equation_ids);
-            global_master_slave_constraint->second->CalculateLocalSystem(mMasterWeightsVector, slave_constant);
-            for (IndexType i_master = 0; i_master < master_equation_ids.size(); ++i_master)
-            {
-                rTransformationMatrixLocal(slave_index, i_masters_total) += mMasterWeightsVector(i_master);
-                i_masters_total++;
-            }
+            global_master_slave_constraint->second->CalculateLocalSystem(master_weights_vector, slave_constant);
+
+            IndexType master_index = 0;
+            double master_weight = 0.0;
+            IndexType i_master = 0;
+            for (auto&  master_eq_id : master_equation_ids)
+            { // Loop over all the masters the slave has
+                master_index = std::distance(rEquationIds.begin(), std::find(rEquationIds.begin(), rEquationIds.end(), master_eq_id));
+                //master_weight = mTransformationMatrixLocal(slave_index,master_index);
+                master_weight = master_weights_vector(i_master);
+                for (auto& internal_index : mLocalIndices.internal_index_vector) {
+                    // For K(m,u) and K(u,m)
+                    rLHSContribution(internal_index, master_index) += rLHSContribution(internal_index, slave_index) * master_weight;
+                    rLHSContribution(master_index, internal_index) += rLHSContribution(slave_index, internal_index) * master_weight;
+                }
+
+                mLocalIndices.container_master_weights.push_back( master_weight );
+                mLocalIndices.container_master_slaves.push_back( slave_index );
+                mLocalIndices.processed_master_indices.push_back( master_index );
+                i_master++;
+            } // Loop over all the masters the slave has
         }
-        for (const auto &master_index : rLocalIndices.master_index_vector)
-            rTransformationMatrixLocal(master_index, master_index) = 1.0;
 
-        for (const auto &internal_index : rLocalIndices.internal_index_vector)
-            rTransformationMatrixLocal(internal_index, internal_index) = 1.0;
+        //Adding contribution from slave to Kmm
+        IndexType master_i = 0;
+        for (auto& master_index : mLocalIndices.processed_master_indices) {
+            IndexType master_i_other = 0;
+            for (auto& master_index_other : mLocalIndices.processed_master_indices) {
+                rLHSContribution(master_index, master_index_other) += mLocalIndices.container_master_weights[master_i] *
+                                                                        rLHSContribution(mLocalIndices.container_master_slaves[master_i], mLocalIndices.container_master_slaves[master_i_other])
```

Comment:
are u accessing to a map here? take care is expensive...

Reply:
These are vectors @RiccardoRossi  not maps.